src/blr/utils.py

```diff
@@ -79,49 +79,68 @@ def print_stats(summary, name=None, value_width=15, print_to=sys.stderr):
     print("="*width, file=print_to)
 
 
-def create_header(file, name):
-    """
-    Create SAM header dict with new tool and command line argument information based on template file. Appends new PG
-    entry with tool name (ID), software name (PN), command line arguments (CL) to track the tools applied to the file.
-    Use in output SAM/BAM file as 'header' attribute.
+class PySAMIO:
+    """ Reader and writer for BAM/SAM files that automatically attaches processing step information to header """
 
-    :param file: string. Path to input SAM/BAM file.
-    :param name: string. Pass '__name__' variable to be used to get program and tool name.
-    :return: pysam.AlignmentHeader object
-    """
-    def check_name(identifier, prev_entries):
+    def __init__(self, inname: str, inmode: str, outname: str, outmode: str, name: str):
         """
-        Check that identifier does not appear in previous header entries. Update with numbered suffix if the name is
-        not unique.
-        :param identifier: string. Program record identifier
-        :param prev_entries: dict. Dictionary of "PG" header entries from pysam.AlignmentFile header.
-        :return: string. Updated identifier
+        :param inname: Path to input SAM/BAM file.
+        :param inmode: Reading mode for input file. 'r' for SAM and 'rb' for BAM.
+        :param outname: Path to output SAM/BAM file.
+        :param outmode: Reading mode for output file. 'r' for SAM and 'rb' for BAM.
```

Comment:
If we're making a class for doing everything automatically I think we could go for automatically detecting format from the file name, otherwise defaulting to BAM (as I see it most commonly used).